STM32Cube/Tasks/trajectory.c

```diff
@@ -0,0 +1,256 @@
+    Forces forces;
+    float angle = atan(velX/velY);
+    float Fd = -interpolate_drag(extension, sqrt(velY*velY+ velX*velX), alt); // force of drag (N)
+    float Fg = -gravitational_acceleration(alt) * mass; // force of gravity (N)
+    forces.Fy = Fd * cos(angle) + Fg;
+    forces.Fx = Fd * sin(angle);
+    return forces;
+}
+
+/**
+    * rk4 method to integrate altitude from velocity, and integrate velocity from acceleration (force/mass)
+    * @param h time step
+    * @param extension extension of airbrakes, 0-1
+    * @param mass of rocket (kg)
+    * @param state, including altitude (m) and velocity in X and Y directions (m/s)
+    * @return updated altitude and velocity integrals after one rk4 step
+*/
+RK4State rk4(float h, float mass, float extension, RK4State state) {
+    Forces forces;
+
+    forces = get_forces(extension, mass, state.velX, state.velY, state.alt);
+    float ka1 = h * state.velY;
+    float kvY1 = h * acceleration(forces.Fy, mass);
+    float kvX1 = h * acceleration(forces.Fx, mass);
+
+    forces = get_forces(extension, mass, state.velX + kvX1/2, state.velY + kvY1/2, state.alt + ka1/2);
+    float ka2 = h * (state.velY + h*kvY1/2);
+    float kvY2 = h * acceleration(forces.Fy, mass);
+    float kvX2 = h * acceleration(forces.Fx, mass);
+
+    forces = get_forces(extension, mass, state.velX + kvX2/2, state.velY + kvY2/2, state.alt + ka2/2);
+    float ka3 = h * (state.velY + h*kvY2/2);
+    float kvY3 = h * acceleration(forces.Fy, mass);
+    float kvX3 = h * acceleration(forces.Fx, mass);
+
+    forces = get_forces(extension, mass, state.velX + kvX3, state.velY + kvY3, state.alt + ka3);
+    float ka4 = h * (state.velY + h*kvY3);
+    float kvY4 = h * acceleration(forces.Fy, mass);
+    float kvX4 = h * acceleration(forces.Fx, mass);
+
+    RK4State updatedState;
+    updatedState.alt = (state.alt + (ka1 + 2*ka2 + 2*ka3 + ka4)/6);
+    updatedState.velY = (state.velY + (kvY1 + 2*kvY2 + 2*kvY3 + kvY4)/6);
+    updatedState.velX = (state.velX + (kvX1 + 2*kvX2 + 2*kvX3 + kvX4)/6);
+
+    return updatedState;
+}
+
+/** @return max apogee
+* @param velocity vertical velocity (m/s)
+* @param altitude (m)
+* @param airbrake_ext extension of airbrakes, 0-1
+* @param mass (kg)
+*/
+float get_max_altitude(float velY, float velX, float altitude, float airbrake_ext, float mass) {
+
+    const float h = 0.05; // interval of change for rk4
+    float prevAlt = 0.0; // variable to store previous altitude
+        
+    RK4State states;
+    
+    states.alt = altitude;
+    states.velY = velY;
+    states.velX = velX;
+
+    while (states.alt >= prevAlt) {
+        prevAlt = states.alt; //to check if altitude is decreasing to exit the loop
+        states = rk4(h, mass, airbrake_ext, states);  // update velocity and altitude
+
+    }
+
+    return prevAlt;
+}
+
+void trajectory_task(){    
+    altQueue = xQueueCreate(1, sizeof(AltTime));
+    angleQueue = xQueueCreate(1, sizeof(AnglesUnion));
```

Comment:
Cannot declare these here, because the angle queue will probably be written to before this task has even begun executing. Same with the altitude.